infrastructure/src/main/java/org/corfudb/infrastructure/LivenessValidator.java

```diff
@@ -0,0 +1,177 @@
+        }
+        return false;
+    }
+
+    private long getHeartbeat(TableName table) {
+        long syncHeartBeat = LIVENESS_INIT_VALUE;
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            ActiveCPStreamMsg activeCPStreamMsg = (ActiveCPStreamMsg) txn.getRecord(
+                    CompactorMetadataTables.ACTIVE_CHECKPOINTS_TABLE_NAME, table).getPayload();
+            txn.commit();
+            if (activeCPStreamMsg != null) {
+                syncHeartBeat = activeCPStreamMsg.getSyncHeartbeat();
+            }
+        } catch (Exception e) {
+            log.warn("Unable to acquire ActiveCPStreamMsg {}", table, e);
+        }
+        return syncHeartBeat;
+    }
+
+    @VisibleForTesting
+    public long getCpStreamTail(TableName table) {
+        String cpStreamName = TableRegistry.getFullyQualifiedTableName(table.getNamespace(), table.getTableName());
+        UUID cpStreamId = CorfuRuntime.getCheckpointStreamIdFromName(cpStreamName);
+        return corfuRuntime.getSequencerView()
+                .getStreamAddressSpace(new StreamAddressRange(cpStreamId, Address.MAX, Address.NON_ADDRESS)).getTail();
+    }
+
+    private int getIdleCount() {
+        int idleCount = 0;
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            idleCount = txn.executeQuery(CompactorMetadataTables.CHECKPOINT_STATUS_TABLE_NAME, entry -> (
+                    (CheckpointingStatus) entry.getPayload()).getStatus() == StatusType.IDLE).size();
+            log.trace("Number of idle tables: {}", idleCount);
+            txn.commit();
+        }
+        return idleCount;
+    }
+
+    public StatusToChange shouldChangeManagerStatus(Duration currentTime) {
+        //Find the number of tables with IDLE status
+        long idleCount = getIdleCount();
+        if (livenessValidatorHelper.getPrevActiveTime().equals(Duration.ofMillis(LIVENESS_INIT_VALUE)) ||
+                idleCount < livenessValidatorHelper.getPrevIdleCount()) {
+            log.trace("Checkpointing in progress...");
+            livenessValidatorHelper.updateValues(idleCount, currentTime);
+            return StatusToChange.NONE;
+        }
+
+        Optional<CheckpointingStatus> managerStatus = Optional.empty();
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            managerStatus = Optional.ofNullable((CheckpointingStatus) txn.getRecord(
+                    CompactorMetadataTables.COMPACTION_MANAGER_TABLE_NAME,
+                    CompactorMetadataTables.COMPACTION_MANAGER_KEY).getPayload());
+            txn.commit();
+        } catch (Exception e) {
+            log.warn("Unable to acquire Manager Status");
+        }
+
+        boolean isTimedOut = currentTime.minus(livenessValidatorHelper.getPrevActiveTime()).compareTo(timeout) > 0;
+
+        if (idleCount == 0 || isTimedOut &&
+                managerStatus.isPresent() && managerStatus.get().getStatus() == StatusType.STARTED) {
```

Comment:
If timeout has passed and it always fails to get a managerStatus, shouldn't it be marked as FINISH and change terminate this cycle?

Reply:
Resolving as per offline discussion